### ISSUE TITLE
Unhandled Exception: SocketException: OS Error on Disconnect.

### ISSUE THREAD

**Issue by JacoFourie** [shamblett/mqtt_client#317]
Hi.

I get this error when I disconnect. I know you don't use Flutter so not sure if this only happens in Flutter.
The app will crash sometimes on iOS if i try to reconnect. Can this Connection reset by peer error be handled?

```
I/flutter ( 3207): 2-2021-09-05 20:15:49.627097 -- MqttServerConnection::_onData - message available event fired
I/flutter ( 3207): 2-2021-09-05 20:15:49.627397 -- MqttConnectionHandlerBase::messageAvailable - message type is MqttMessageType.publish
I/flutter ( 3207): 2-2021-09-05 20:15:49.627616 -- PublishingManager::handlePublish - publish received from broker with topic Inverter/aes_data_comp
I/flutter ( 3207): 2-2021-09-05 20:15:49.627825 -- MqttConnectionHandlerBase::sendMessage - MQTTMessage of type MqttMessageType.publishAck
I/flutter ( 3207): Header: MessageType = MqttMessageType.publishAck, Duplicate = false, Retain = false, Qos = MqttQos.atMostOnce, Size = 0
I/flutter ( 3207): PublishAck Variable Header: MessageIdentifier={3}
I/flutter ( 3207): 2-2021-09-05 20:15:50.235066 -- MqttConnectionHandlerBase::disconnect - entered
I/flutter ( 3207): 2-2021-09-05 20:15:50.236180 -- MqttConnectionHandlerBase::sendMessage - MQTTMessage of type MqttMessageType.disconnect
I/flutter ( 3207): Header: MessageType = MqttMessageType.disconnect, Duplicate = false, Retain = false, Qos = MqttQos.atMostOnce, Size = 0
I/flutter ( 3207): 2-2021-09-05 20:15:50.236990 -- MqttConnectionHandlerBase::_performConnectionDisconnect entered
I/flutter ( 3207): 2-2021-09-05 20:15:50.239047 -- MqttConnectionKeepAlive::stop - stopping keep alive
E/flutter ( 3207): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: SocketException: OS Error: Connection reset by peer, errno = 104, address = x.x.x.x, port = 55434
E/flutter ( 3207): 
```

**Comment by JacoFourie** [comment]
OK so it seems if I add Try Catch around the code all is OK.

**Comment by JacoFourie** [comment]
Nope spoke to soon. It is still crashing.

**Comment by shamblett** [comment]
This has cropped up before, if your try catch block isn't catching the exception at all you may need to ask the Google flutter devs why. You say only on IOS, does the exception get raised/handles OK on Android?

**Comment by JacoFourie** [comment]
I do not get any crashes on Android and it seems to work 100% in the iOS simulator but will crash randomly on the iOS device.

**Comment by JacoFourie** [comment]
OK I managed to get this now in the IDE

```
#0      _NativeSocket.write (dart:io-patch/socket_patch.dart:1114:34)
#1      _RawSocket.write (dart:io-patch/socket_patch.dart:1759:15)
#2      _Socket._write (dart:io-patch/socket_patch.dart:2192:18)
#3      _SocketStreamConsumer.write (dart:io-patch/socket_patch.dart:1940:26)
#4      _SocketStreamConsumer.addStream.<anonymous closure> (dart:io-patch/socket_patch.dart:1914:11)
#5      _rootRunUnary (dart:async/zone.dart:1362:47)
#6      _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#7      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#8      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#9      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#10     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#11     _StreamController._add (dart:asy<…>
Lost connection to device.
```

**Comment by shamblett** [comment]
OK but the above stack trace is all in the Dart runtime, not the client, what do you propose the client do about this? As I've said  above you are better of asking the Google guys why IOS does this on occasion.

**Comment by JacoFourie** [comment]
I just though I would give you feedback on what I saw.

**Comment by shamblett** [comment]
Yes maybe of use to other IOS users, there should really be a wiki for this stuff somewhere, there' s quite a lot of things like this cropped up over the years, good info to have but not an issue with the client as such. I'll have a look at what can be done here.

**Comment by JacoFourie** [comment]
I hope this is the case but I think the issue has been resolved after I upgraded to Flutter 2.5. I have not had a random crash since then.
